components/packages/index.rst

```diff
@@ -0,0 +1,259 @@
+configuration modularization. You'll likely want to break your configuration into common (groups of) elements, building
+it into reusable pieces which can subsequently be used by many/all devices. Only unique pieces of your configuration
+remain in any given device's YAML configuration file.
+
+This can be accomplished with ESPHome's "packages" feature.
+
+All definitions from packages will be merged with your device's main configuration in a non-destructive way. This
+allows overriding (parts of) configuration contained in the package(s). Substitutions in your main configuration will
+override substitutions with the same name in a package.
+
+Dictionaries are merged key-by-key. Lists of components are merged by component ID (if specified). Other lists are
+merged by concatenation. All other configuration values are replaced with the later value.
+
+Local Packages
+--------------
+
+Consider the following example where the author put common pieces of configuration (like Wi-Fi and I²C) into base files
+and then extends it with some device-specific configuration in the main configuration.
+
+Note how the piece of configuration describing ``api`` component in ``device_base.yaml`` gets merged with the services
+definitions from main configuration file.
+
+.. code-block:: yaml
+
+    # In config.yaml
+    substitutions:
+      node_name: mydevice
+      device_verbose_name: "My Device"
+
+    packages:
+      wifi: !include common/wifi.yaml
+      device_base: !include common/device_base.yaml
+
+    api:
+      services:
+        - service: start_laundry
+          then:
+            - switch.turn_on: relay
+            - delay: 3h
+            - switch.turn_off: relay
+
+    sensor:
+      - platform: mhz19
+        co2:
+          name: "CO2"
+        temperature:
+          name: "Temperature"
+        update_interval: 60s
+        automatic_baseline_calibration: false
+
+.. code-block:: yaml
+
+    # In wifi.yaml
+    wifi:
+      ssid: !secret wifi_ssid
+      password: !secret wifi_password
+      domain: .yourdomain.lan
+      fast_connect: true
+
+.. code-block:: yaml
+
+    # In device_base.yaml
+    esphome:
+      name: ${node_name}
+      platform: ESP32
+      board: wemos_d1_mini32
+      build_path: ./build/${node_name}
+
+    # I²C Bus
+    i2c:
+      sda: GPIOXX
+      scl: GPIOXX
+      scan: true
+      frequency: 100kHz
+
+    # Enable logging
+    logger:
+      level: ${log_level}
+
+    api:
+      encryption:
+        key: !secret api_encryption_key
+      reboot_timeout: 1h
+
+    sensor:
+      - <<: !include common/sensor/uptime.config.yaml
+      - <<: !include common/sensor/wifi_signal.config.yaml
+    binary_sensor:
+      - <<: !include common/binary_sensor/connection_status.config.yaml
+
+    switch:
+      - <<: !include common/switch/restart_switch.config.yaml
+
+.. _config-git_packages:
+
+Remote/Git Packages
```

Comment:
Should have a `Configuration variables` section for this one as it has a specific schema.